packages/core/src/shared/fs/fs.ts

```diff
@@ -211,6 +222,37 @@ export class FileSystem {
             return nodefs.rename(oldUri.fsPath, newUri.fsPath).catch(errHandler)
         }
 
+        /**
+         * We were seeing 'FileNotFound' errors during renames, even though we did a `writeFile()` right before the rename.
+         * The error looks to be from here: https://github.com/microsoft/vscode/blob/09d5f4efc5089ce2fc5c8f6aeb51d728d7f4e758/src/vs/platform/files/node/diskFileSystemProvider.ts#L747
+         * So a guess is that the exists()(stat() under the hood) call needs to be retried since there may be a race condition.
+         */
+        let attempts = 0
+        const isExists = await waitUntil(async () => {
+            const result = await fs.exists(oldUri)
+            attempts += 1
+            return result
+        }, renameTimeoutOpts)
+        // TODO: Move the `aws_fileSystem` or some variation of this metric in to common telemetry
+        if (!isExists) {
+            // source path never existed after multiple attempts.
+            // Either the file never existed, or had we waited longer it would have. We won't know.
+            telemetry.aws_fileSystem.emit({
+                result: 'Failed',
+                action: 'rename',
+                reason: 'SourceNotExists',
+                reasonDesc: `After ${renameTimeoutOpts.timeout}ms the source path did not exist: ${oldUri.fsPath}`,
+            })
+        } else if (attempts > 1) {
+            // Indicates that rename() would have failed if we had not waited for it to exist.
+            telemetry.aws_fileSystem.emit({
+                result: 'Succeeded',
+                action: 'rename',
+                reason: 'RenameRaceCondition',
+                reasonDesc: `After ${attempts} attempts the source path existed: ${oldUri.fsPath}`,
```

Comment:
there is an `attempts` field: https://github.com/aws/aws-toolkit-common/blob/2ca7102873d5c9a1ad166cbcc6fa127d22b461c7/telemetry/definitions/commonDefinitions.json#L101

should we use fields instead of using a human-readable message to store data